Arch arm sw vector table cleanup

ioannisg opened the pull request:
Now in case of boot-able application CONFIG_SW_VECTOR_RELAY_CLIENT is used to setting the
vector table pointer in RAM so the parent image can forward the
interrupts to it

This will save some flash on Cortex-M0 with SW vector relaying, and in particular, (16 + 32)x4 = 192 bytes.

ioannisg committed:
arch: arm: cortex_m: Add config for SW_VECTOR_RELAY_CLIENT

Define vector relay tables for bootloader only.
If an image is not a bootloader image (such as an MCUboot image)
but it is a standard Zephyr firmware, chain-loadable by a
bootloader, then this image will not need to relay IRQs itself.
In this case SW_VECTOR_RELAY_CLIENT should be used to setting the
vector table pointer in RAM so the parent image can forward the
interrupts to it.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Co-authored-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>
Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

stephanosio commented on arch/arm/core/aarch32/cortex_m/CMakeLists.txt:
```diff
@@ -13,12 +13,18 @@ zephyr_library_sources(
 
+# When CONFIG_BOOTLOADER_MCUBOOT is set the firmware image
+# is chain-loadable and does not need to forward interrupts.
+if(!CONFIG_BOOTLOADER_MCUBOOT)
```
```suggestion
if(NOT CONFIG_BOOTLOADER_MCUBOOT)
```

stephanosio commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -270,8 +270,13 @@ config SW_VECTOR_RELAY
+	  building a chain-loadable firmware, this option instructs
+	  the kernel to store a pointer to the vector table onto an
+	  address in RAM so it can be used by the bootloader fimrware
+	  yo implement the software vector relaying. This is only
```
```suggestion
	  you implement the software vector relaying. This is only
```

ioannisg replied:
Actually, it should be to* , but thanks for spotting it :)

stephanosio commented on arch/arm/core/aarch32/cortex_m/CMakeLists.txt:
```diff
@@ -13,12 +13,18 @@ zephyr_library_sources(
+# is chain-loadable and does not need to forward interrupts.
+if(!CONFIG_BOOTLOADER_MCUBOOT)
+
 zephyr_linker_sources_ifdef(CONFIG_SW_VECTOR_RELAY
   ROM_START
   SORT_KEY 0x0relay_vectors
   relay_vector_table.ld
   )
```
need to indent this

```suggestion
  zephyr_linker_sources_ifdef(CONFIG_SW_VECTOR_RELAY
    ROM_START
    SORT_KEY 0x0relay_vectors
    relay_vector_table.ld
    )
```

stephanosio reviewed (commented):
Some nits

stephanosio commented on arch/arm/core/aarch32/irq_relay.S:
```diff
@@ -26,7 +26,10 @@
+/* When CONFIG_BOOTLOADER_MCUBOOT is set the firmware image
+ * is chain-loadable and does not need to forward interrupts.
+ */
+#if defined(CONFIG_SW_VECTOR_RELAY) && !defined(CONFIG_BOOTLOADER_MCUBOOT)
```
I wonder if this would be better encoded in Kconfig.

oyvindronningstad replied:
I agree. There could be a concept of VECTOR_RELAY_CLIENT or similar, which just allocates the RAM word for the pointer.

carlescufi replied:
But in that case (being a chain-loadable image) why would you enable `SW_VECTOR_RELAY` at all?

oyvindronningstad replied:
You wouldn't. You would enable SW_VECTOR_RELAY_CLIENT instead

ioannisg replied:
Yeap, @oyvindronningstad is correct here

nvlsianpu commented:
good - have to test this.

ioannisg commented:
> good - have to test this.

Yeap, please help with some testing here, @nvlsianpu , you master this area :)

nvlsianpu commented:
@ioannisg Work just fine (with @stephanos corrections). Tested on nrf51dk_nrf51244.

``` console
west build -d build/mcuboot_51 -b nrf51dk_nrf51422 ./bootloader/mcuboot/boot/zephyr/ -t menuconfig
/* ^^ have selected LOG */
west build -d build/mcuboot_51 -b nrf51dk_nrf51422 ./bootloader/mcuboot/boot/zephyr/
west flash -d build/mcuboot_51
west build -d build/ble -b nrf51dk_nrf51422 ./zephyr/samples/bluetooth/beacon/ -t menuconfig
/* ^^I  have selected BOOTLOADER_MCUBOOT and SW_VECTOR_RELAY via menuconfig. Normally In proj.conf BOOTLOADER_MCUBOOT set is enought */
west build -d build/ble -b nrf51dk_nrf51422 ./zephyr/samples/bluetooth/beacon/
sign_img 0x1a000 build/ble/zephyr/zephyr.hex  signed_ble_51.hex
nrfjprog -f NRF51 --program signed_ble_51.hex
nrfjprog -f NRF51 -r
```

console output:
``` console
*** Booting Zephyr OS build zephyr-v2.3.0-438-g24f8b5888a34  ***
[00:00:00.006,500] <inf> mcuboot: Starting bootloader
[00:00:00.017,181] <inf> mcuboot: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00:00:00.030,151] <inf> mcuboot: Boot source: none
[00:00:00.037,689] <inf> mcuboot: Swap type: none
[00:00:01.432,617] <inf> mcuboot: Bootloader chainload address offset: 0x8000
[00:00:01.441,864] <inf> mcuboot: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v2.3.0-438-g24f8b5888a34  ***
Starting Beacon Demo
Bluetooth initialized
Beacon started
[00:00:00.012,969] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
[00:00:00.013,000] <inf> bt_hci_core: HW Variant: nRF51x (0x0001)
[00:00:00.013,031] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.3 Build 99
[00:00:00.015,686] <inf> bt_hci_core: Identity: e1:f7:d2:4a:e3:b6 (random)
[00:00:00.015,716] <inf> bt_hci_core: HCI: version 5.2 (0x0b) revision 0x0000, manufacturer 0x05f1
[00:00:00.015,747] <inf> bt_hci_core: LMP: version 5.2 (0x0b) subver 0xffff
```
`Test beacon` appeared on my BLE scaner as well.

nvlsianpu reviewed (commented):
Patch within correction suggested is correct.

oyvindronningstad reviewed (changes requested):
Please consider encapsulating it in a new kconfig, as mentioned by @stephanosio. This will also help in integration with out-of-tree bootloaders.

nvlsianpu commented:
@oyvindronningstad The idea is not clear too me - can you elaborate more or even better give a code snippet?

oyvindronningstad commented:
> @oyvindronningstad The idea is not clear too me - can you elaborate more or even better give a code snippet?

https://github.com/oyvindronningstad/zephyr/commit/ca34f8ebae472228c15dd1b6c6cf81280fb37adc

nvlsianpu commented:
@stephanosio @ioannisg @oyvindronningstad 
I put on Øyvind's sugestion - the last commit should be squashed with the previous one.

nvlsianpu commented:
Patch squashed to one commit (code is same as @oyvindronningstad approved).
@stephanosio @d3zd3z Would be great to have your review here.

nvlsianpu committed:
Kconfig: update BOOTLOADER_MCUBOOT help message

Description was updated so reflects fact that this
option activates SW_VECTOR_RELAY_CLIENT feature now.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
